File: beecrowd/3006 - Passeio em Copacabana.cpp
#include <bits/stdc++.h>

using namespace std;

const int N = 1e6 + 5, K = 4;

int dist[N][K];

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    size_t t;
    char store[] = {'C', 'F'};

    cin >> t;

    for(size_t i = 0; i < t; ++i) {
        string avenue;
        queue<pair<int, int>> q;

        cin >> avenue;

        int n = avenue.size();

        for(int i = 0; i < n; ++i)
            for(int j = 0; j < K; ++j)
                dist[i][j] = -1;

        for(int i = 0; i < n; ++i) {
            if(avenue[i] == 'R') {
                dist[i][0] = 0;
                q.emplace(i, 0);
            }
        }

        optional<int> min_dist;

        while(!q.empty()) {
            auto [x, mask] = q.front();

            q.pop();

            if(mask == 3) {
                min_dist = dist[x][mask];
                break;
            }

            for(int delta : {-1, 1}) {
                int dest = x + delta;
                
                if(dest < 0 || dest >= n)
                    continue;

                int new_mask = mask;

                for(int i = 0; i < 2; ++i) {
                    if(avenue[dest] == store[i])
                        new_mask |= 1 << i;
                }

                if(dist[dest][new_mask] != -1)
                    continue;

                dist[dest][new_mask] = 1 + dist[x][mask];
                q.emplace(dest, new_mask);
            }
        }

        assert(static_cast<bool>(min_dist));
        cout << *min_dist << '\n';
    }

    return 0;
}

File: beecrowd/3006 - Passeio em Copacabana.rs
#![allow(dead_code)]
use std::{
    collections::VecDeque,
    io::{self, Read, Write},
    writeln,
};

struct Scanner {
    buffer: Vec<u8>,
    index: usize,
}

impl Scanner {
    fn new() -> Self {
        let mut input = Vec::new();

        io::stdin().read_to_end(&mut input).unwrap();

        Self {
            buffer: input,
            index: 0,
        }
    }

    fn has_next(&mut self) -> bool {
        while self.index < self.buffer.len() && self.buffer[self.index].is_ascii_whitespace() {
            self.index += 1;
        }

        self.index < self.buffer.len()
    }

    fn next<T: std::str::FromStr>(&mut self) -> T {
        if !self.has_next() {
            panic!("End Of File");
        }

        let start = self.index;

        while self.index < self.buffer.len() && !self.buffer[self.index].is_ascii_whitespace() {
            self.index += 1;
        }

        std::str::from_utf8(&self.buffer[start..self.index])
            .unwrap()
            .parse::<T>()
            .ok()
            .unwrap()
    }
}

fn main() {
    let mut writer = io::BufWriter::with_capacity(1 << 20, io::stdout());

    let mut sc = Scanner::new();

    let t = sc.next::<usize>();

    let store = ['C', 'F'];
    let delta = [-1, 1];

    for _ in 0..t {
        let avenue = sc.next::<String>();

        let n = avenue.len();

        let mut dist = vec![[-1; 4]; n];
        let mut queue = VecDeque::new();

        let chars = avenue.chars().collect::<Vec<_>>();

        for (i, c) in chars.iter().enumerate() {
            if *c == 'R' {
                dist[i][0] = 0;
                queue.push_back((i, 0));
            }
        }

        let mut answer = None;

        while let Some((x, mask)) = queue.pop_front() {
            if mask == 3 {
                answer = Some(dist[x][mask]);
                break;
            }

            for &dx in &delta {
                let dest = x as i32 + dx;

                if dest < 0 || dest >= n as i32 {
                    continue;
                }

                let mut new_mask = mask;

                for (i, &c) in store.iter().enumerate() {
                    if chars[dest as usize] == c {
                        new_mask |= 1 << i;
                        break;
                    }
                }

                if dist[dest as usize][new_mask] != -1 {
                    continue;
                }

                dist[dest as usize][new_mask] = 1 + dist[x][mask];
                queue.push_back((dest as usize, new_mask));
            }
        }

        assert!(answer.is_some());

        writeln!(writer, "{}", answer.unwrap()).ok();
    }
}
